### DISCUSSION TITLE
Variable scaling

### DISCUSSION THREAD

**Discussion by Edward-Eth** [idaholab/moose#21297] [Q&A General] [answered]
I am running a coupled physics simulation, and having some convergence issues. These are my per-variable residuals at the start and end of a converged timestep:

Start:
disp_x: 0.0207124		
disp_y: 0.0538449		
disp_z: 0.0559112		
T:      1213.58			

End:
disp_x: 1.03655e-07
disp_y: 6.19961e-08
disp_z: 2.02929e-07
T:      2.23613e-12

According to the "Troubleshooting failed solves" page, my residuals should be within 1 order of magnitude of each other. These obviously aren't, but at the start of the step (initial guess) T is much larger, and at the end it is much smaller. Which do I scale for? Should I multiply T by ~10^4 so that when converged it is a similar magnitude to the other variables, or should I times disp_x,y,z, by ~10^5 to bring initial guess in line with each other?

Additionally, here are the residuals for a failed time step:

Initial:
disp_x: 1.92706e-08
disp_y: 8.08506e-09
disp_z: 1.65146e-08
T:      54.5449

End:
disp_x: 4.41008e-05
disp_y: 2.61794e-05
disp_z: 4.3757e-05
T:      5.39617e-12

Again. at the start the T contribution is much larger, but by the end (after 20 NL iterations and failing to converge), T is much smaller.

**Comment by hugary1995** [comment, marked as answer]
You could try automatic scaling which scales the variables such that the Jacobian is close to unity in sup-norm. If you were using Newton and non-AD kernels, the results suggest that the displacements and temperature converge at different rates, and the Jacobian for the displacement subproblem (or its coupling with temperature) might be off. You could try the new mechanics system which should provide you with a perfect Jacobian. 

Anyways, automatic scaling (or maybe reference residual problem) + Newton + perfect Jacobian should resolve this issue.

**Reply by Edward-Eth** [reply]
Hi, I'm already using AD and automatic scaling here.

**Reply by Edward-Eth** [reply]
More specifically, I am using newton solve, AD for TensorMechanics and all relevant material properties, AD for Heat Conduction and Heat Conduction time derivative (and all relevant materials properties), and am coupling using AD compute thermal expansion eigenstrain.

**Reply by hugary1995** [reply]
Okay, I would then check if everything in your input file is AD, do you mind pasting the input file here? If the jacobian is perfect and you are taking more than 20 nl iterations as you mentioned, the time step you are taking might be too large that the initial guess is outside the basin of a traction.

**Reply by Edward-Eth** [reply]
Not sure I will be able to post the input file, but all of my BCs, materials and kernels are AD. The only things not AD are an element activator user object,  a property read file user object, and functions, none of which can be AD to the best of my knowledge.

**Reply by Edward-Eth** [reply]
As for the time step, that is after it is cut down to less than 1 second, which should definitely be stable :/

**Reply by hugary1995** [reply]
This isn't a scaling issue, it's a wrong jacobian issue.

**Reply by Edward-Eth** [reply]
Is there any way to track down where the incorrect jacobian contributions are coming from? The only "loading" in the structural/tensor mechanics part of the model is from thermal expansion, so is that the most likely source of the issue?

**Reply by Edward-Eth** [reply]
Am I correct in thinking that if the issue is with the Jacobian, then a Jacobian free solve (PJFNK) will in theory not be affected? Running one now to see anyway but I would be interested in your thoughts on that approach.

**Reply by hugary1995** [reply]
First, if you can manage to reduce the size of the problem to a few elements, you'll be able to check the accuracy of the jacobian by passing the command line argument `-snes_test_jacobian`, which can give us some idea about the quality of the jacobian.

Next, you'll need to tell us whether the non AD objects introduce material properties that depend on nonlinear variables. If so, that's likely gonna be the source of inaccurate jacobian.

**Reply by GiudGiud** [reply]
A JFNK solve does not depend on the Jacobian. A PJFNK solve will use the Jacobian to precondition the system, so if the Jacobian is wrong it will still impact the solve. 

Gary is right that you should check your implementation at this point. Use the jacobian tester, and verify that you are consistently using AD or have implemented Diagonal and off-diagonal contributions to the Jacobian correctly

**Comment by GiudGiud** [comment]
Hello

so you could try manually scaling the equations, but you ll have to use the scaling factors appropriate for the final state not the initial state. It seems the compared scale of the residuals is inverted between the two.

automatic_scaling is a good idea too. I think this is one of the cases where you will want to update the scaling factors.
so add `compute_scaling_once = false` to the Executioner block.

Guillaume

**Comment by mangerij** [comment]
Another suggestion: sometimes going to a different unit system can help convergence. It usually doesn't make sense to have materials coefficients of one part of the coupled problem that are like 1e-15 and then materials coefficients of another part of the coupled problem that are, for example 1e15.
